File: apps/bible/android/app/src/main/kotlin/app/luxbible/app/VerseOfTheDayWidgetStore.kt
package app.luxbible.app

import android.content.Context
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/** A single day's Verse of the Day, written by the Flutter app and read by the widget. */
data class VerseOfTheDayWidgetEntry(
    /** The local calendar date this verse belongs to, formatted as `yyyy-MM-dd`. */
    val date: String,
    val reference: String,
    val translation: String,
    val text: String,
)

/**
 * The horizon of verses the app last pushed, held as the JSON both platforms share.
 *
 * The app owns every user-facing string in it, so the widget never has to resolve a passage, a
 * translation fallback, or a book name of its own.
 */
object VerseOfTheDayWidgetStore {
    private const val PREFERENCES = "verse_of_the_day_widget"
    private const val PAYLOAD_KEY = "verseOfTheDay"

    fun write(context: Context, json: String) {
        context.preferences.edit().putString(PAYLOAD_KEY, json).apply()
    }

    /**
     * The entry for the local day [date] falls on, or null when the horizon does not reach it.
     *
     * Every field is required, so a payload left behind by an older install whose shape has since
     * changed reads as nothing rather than as blanks.
     */
    fun read(context: Context, date: Date): VerseOfTheDayWidgetEntry? {
        val json = context.preferences.getString(PAYLOAD_KEY, null) ?: return null
        val isoDate = SimpleDateFormat("yyyy-MM-dd", Locale.US).format(date)

        return runCatching {
            val entries = JSONObject(json).getJSONArray("entries")
            (0 until entries.length())
                .map { entries.getJSONObject(it) }
                .firstOrNull { it.getString("date") == isoDate }
                ?.let {
                    VerseOfTheDayWidgetEntry(
                        date = isoDate,
                        reference = it.getString("reference"),
                        translation = it.getString("translation"),
                        text = it.getString("text"),
                    )
                }
        }.getOrNull()
    }

    private val Context.preferences
        get() = getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
}

File: apps/bible/ios/VerseOfTheDayWidget/Shared/VerseOfTheDayWidgetStore.swift
import Foundation

// The app resolves every user-facing string so the widget never has to.
struct VerseOfTheDayWidgetEntry: Codable, Equatable {
  let date: String
  let reference: String
  let translation: String
  let text: String
}

struct VerseOfTheDayWidgetPayload: Codable, Equatable {
  let entries: [VerseOfTheDayWidgetEntry]
}

enum VerseOfTheDayWidgetStore {
  static let appGroupIdentifier = "group.app.luxbible.app"
  static let payloadKey = "verseOfTheDay"

  // Computed per access so a time zone change mid-process can't leave stale state behind.
  static var calendar: Calendar {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = .current
    return calendar
  }

  private static var dateFormatter: DateFormatter {
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.timeZone = .current
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter
  }

  private static var defaults: UserDefaults? {
    UserDefaults(suiteName: appGroupIdentifier)
  }

  static func write(_ json: String) {
    defaults?.set(json, forKey: payloadKey)
  }

  // Every field is required, so a payload from an older install fails to decode instead of half-reading.
  static func read() -> [VerseOfTheDayWidgetEntry] {
    guard let json = defaults?.string(forKey: payloadKey),
      let data = json.data(using: .utf8),
      let payload = try? JSONDecoder().decode(VerseOfTheDayWidgetPayload.self, from: data)
    else { return [] }

    return payload.entries
  }

  static func startOfDay(for entry: VerseOfTheDayWidgetEntry) -> Date? {
    dateFormatter.date(from: entry.date).map { calendar.startOfDay(for: $0) }
  }

  static func string(from date: Date) -> String {
    dateFormatter.string(from: date)
  }
}
